lib/src/lifecycle_module.dart

```diff
@@ -725,25 +735,36 @@ abstract class LifecycleModule extends SimpleModule
 
       ShouldUnloadResult shouldUnloadResult = shouldUnload();
       if (!shouldUnloadResult.shouldUnload) {
+        _state = _previousState;
+        _previousState = null;
+        _transition = null;
         // reject with shouldUnload messages
         throw new ModuleUnloadCanceledException(
             shouldUnloadResult.messagesAsString());
       }
-
       _willUnloadController.add(this);
-      await Future.wait(_childModules.map((child) => child.unload()));
+      List<Future<Null>> childUnloadFutures = <Future<Null>>[];
+      for (var child in _childModules.toList()) {
+        childUnloadFutures.add(child.unload());
+      }
+      _childModules.clear();
+      await Future.wait(childUnloadFutures);
       await onUnload();
       await _disposableProxy.dispose();
       if (_state == LifecycleState.unloading) {
         _state = LifecycleState.unloaded;
+        _previousState = null;
         _transition = null;
       }
       _didUnloadController.add(this);
-    } catch (error, stackTrace) {
+      await _didUnloadController.close();
+    } on ModuleUnloadCanceledException catch (error, stackTrace) {
       _didUnloadController.addError(error, stackTrace);
```

Comment:
Agreed. My only concern is that someone is relying on this behavior since, I think, it predates this PR. I'm cool with calling it a bug, though, which (I think) allows us to change it with a patch or minor release.